packages/walletconnect-v2/src/index.ts

```diff
@@ -51,22 +64,34 @@ export class WalletConnect extends Connector {
   private readonly options: Omit<WalletConnectOptions, 'rpcMap' | 'chains'>
 
   private readonly rpcMap?: Record<number, string | string[]>
-  private readonly chains: number[]
+  private readonly chains: number[] | ArrayOneOrMore<number> | undefined
+  private readonly optionalChains: number[] | ArrayOneOrMore<number> | undefined
   private readonly defaultChainId?: number
   private readonly timeout: number
 
   private eagerConnection?: Promise<WalletConnectProvider>
 
-  constructor({ actions, options, defaultChainId, timeout = DEFAULT_TIMEOUT, onError }: WalletConnectConstructorArgs) {
+  constructor({ actions, defaultChainId, options, timeout = DEFAULT_TIMEOUT, onError }: WalletConnectConstructorArgs) {
     super(actions, onError)
 
-    const { rpcMap, rpc, chains, ...rest } = options
+    const { rpcMap, rpc, ...rest } = options
 
     this.options = rest
-    this.chains = chains
     this.defaultChainId = defaultChainId
     this.rpcMap = rpcMap || rpc
     this.timeout = timeout
+
+    const { chains, optionalChains } = this.getChainProps(rest.chains, rest.optionalChains, defaultChainId)
+    this.chains = chains
+    this.optionalChains = optionalChains
+  }
+
+  private handleProviderEvents(provider: WalletConnectProvider): WalletConnectProvider {
+    return provider
+      .on('disconnect', this.disconnectListener)
+      .on('chainChanged', this.chainChangedListener)
+      .on('accountsChanged', this.accountsChangedListener)
+      .on('display_uri', this.URIListener)
```

Comment:
This is only done once. I'd rather it just be inlined into `initializeProvider`, but I won't make a stink about it.

Reply:
Yup, this is somewhat related to my previous comment https://github.com/Uniswap/web3-react/pull/868#discussion_r1284651907 and also +1 here. We should definitely keep things inline unless required by this PR.